sklearn/neural_network/_base.py

```diff
@@ -0,0 +1,228 @@
+"""Utilities for the neural network modules
+"""
+
+# Author: Issam H. Laradji <issam.laradji@gmail.com>
+# Licence: BSD 3 clause
+
+import numpy as np
+
+from ..utils.fixes import expit as logistic_sigmoid
+
+
+def identity(X):
+    """Simply return the input array.
+
+    Parameters
+    ----------
+    X : {array-like, sparse matrix}, shape (n_samples, n_features)
+        Data, where n_samples is the number of samples
+        and n_features is the number of features.
+
+    Returns
+    -------
+    X : {array-like, sparse matrix}, shape (n_samples, n_features)
+        Same as the input data.
+    """
+    return X
+
+
+def logistic(X):
+    """Compute the logistic function inplace.
+
+    Parameters
+    ----------
+    X : {array-like, sparse matrix}, shape (n_samples, n_features)
+        The input data.
+
+    Returns
+    -------
+    X_new : {array-like, sparse matrix}, shape (n_samples, n_features)
+        The transformed data.
+    """
+    return logistic_sigmoid(X, out=X)
+
+
+def tanh(X):
+    """Compute the hyperbolic tan function inplace.
+
+    Parameters
+    ----------
+    X : {array-like, sparse matrix}, shape (n_samples, n_features)
+        The input data.
+
+    Returns
+    -------
+    X_new : {array-like, sparse matrix}, shape (n_samples, n_features)
+        The transformed data.
+    """
+    return np.tanh(X, out=X)
+
+
+def relu(X):
+    """Compute the rectified linear unit function inplace.
+
+    Parameters
+    ----------
+    X : {array-like, sparse matrix}, shape (n_samples, n_features)
+        The input data.
+
+    Returns
+    -------
+    X_new : {array-like, sparse matrix}, shape (n_samples, n_features)
+        The transformed data.
+    """
+    np.clip(X, 0, np.finfo(X.dtype).max, out=X)
+    return X
+
+
+def softmax(X):
+    """Compute the  K-way softmax function inplace.
+
+    Parameters
+    ----------
+    X : {array-like, sparse matrix}, shape (n_samples, n_features)
+        The input data.
+
+    Returns
+    -------
+    X_new : {array-like, sparse matrix}, shape (n_samples, n_features)
+        The transformed data.
+    """
+    tmp = X - X.max(axis=1)[:, np.newaxis]
+    np.exp(tmp, out=X)
+    X /= X.sum(axis=1)[:, np.newaxis]
+
+    return X
+
+
+ACTIVATIONS = {'identity': identity, 'tanh': tanh, 'logistic': logistic,
+               'relu': relu, 'softmax': softmax}
+
+
+def logistic_derivative(Z):
```

Comment:
I see. Though it returns value that's derivative of the logistic function with given output value, it's not really taking derivative. It's common for those familiar with neural net, but might be non-obvious to other. 

I have updated the docstring. Do you think this helps to explain the purpose of these function? If you think it's necessary to rename them, any suggestions for better name? Thanks